File: latest_ran_output/gpt_5_4_mini/third_002_naive.py
from flask import Flask, request, jsonify
import os
import hmac
import hashlib
import json
from datetime import datetime, timezone

app = Flask(__name__)

# In a real app, replace this with your database access layer.
# Example structure:
# ORDERS = {
#     "order_123": {"status": "pending", "paid_at": None, "updated_at": None}
# }
ORDERS = {}


WEBHOOK_SECRET = os.environ.get("PAYMENT_WEBHOOK_SECRET", "change-me")


def verify_webhook_signature(raw_body: bytes, signature_header: str) -> bool:
    """
    Verifies HMAC-SHA256 signature from the payment provider.

    Expected header format:
        X-Payment-Signature: sha256=<hex_digest>
    """
    if not signature_header:
        return False

    try:
        prefix, received_sig = signature_header.split("=", 1)
    except ValueError:
        return False

    if prefix.lower() != "sha256" or not received_sig:
        return False

    computed_sig = hmac.new(
        WEBHOOK_SECRET.encode("utf-8"),
        raw_body,
        hashlib.sha256
    ).hexdigest()

    return hmac.compare_digest(computed_sig, received_sig)


@app.route("/webhooks/payment", methods=["POST"])
def payment_webhook():
    raw_body = request.get_data()  # keep raw bytes for signature verification
    signature = request.headers.get("X-Payment-Signature", "")

    if not verify_webhook_signature(raw_body, signature):
        return jsonify({"error": "invalid signature"}), 400

    try:
        payload = json.loads(raw_body.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError):
        return jsonify({"error": "invalid JSON"}), 400

    event_type = payload.get("event")
    data = payload.get("data", {})
    order_id = data.get("order_id")

    if not event_type or not order_id:
        return jsonify({"error": "missing event or order_id"}), 400

    # Ensure order exists; in a real app, query the DB.
    order = ORDERS.get(order_id)
    if order is None:
        # If your provider retries on non-2xx responses, you may prefer 200 here
        # to avoid repeated retries for unknown orders. Adjust per your needs.
        return jsonify({"error": "order not found"}), 404

    now = datetime.now(timezone.utc).isoformat()

    if event_type in ("payment.succeeded", "charge.succeeded", "order.paid"):
        order["status"] = "paid"
        order["paid_at"] = data.get("paid_at", now)
        order["updated_at"] = now
    elif event_type in ("payment.failed", "charge.failed"):
        order["status"] = "payment_failed"
        order["updated_at"] = now
    elif event_type in ("payment.refunded", "charge.refunded"):
        order["status"] = "refunded"
        order["updated_at"] = now
    else:
        # Ignore unhandled events, but acknowledge receipt.
        return jsonify({"received": True, "ignored": True}), 200

    # In a real application, persist the updated order to the database here.
    ORDERS[order_id] = order

    return jsonify({
        "received": True,
        "order_id": order_id,
        "status": order["status"],
    }), 200


if __name__ == "__main__":
    # Example seed data for local testing.
    ORDERS["order_123"] = {"status": "pending", "paid_at": None, "updated_at": None}
    app.run(debug=True)

File: latest_ran_output/gpt_5_4_mini/third_002_naive.js
const express = require('express');
const crypto = require('crypto');
const router = express.Router();

// If you use raw body verification for signature checks, mount this route with express.raw()
// e.g. app.post('/webhooks/payment', express.raw({ type: 'application/json' }), webhookHandler)

const WEBHOOK_SECRET = process.env.PAYMENT_WEBHOOK_SECRET || '';

/**
 * Placeholder order update function.
 * Replace with your database logic (e.g., MongoDB, PostgreSQL, Prisma, Sequelize, etc.).
 */
async function updateOrderStatus(orderId, status, payload) {
  // Example:
  // await Order.updateOne({ _id: orderId }, { $set: { status, webhookPayload: payload, updatedAt: new Date() } });
  console.log(`Updating order ${orderId} to status ${status}`);
  return true;
}

/**
 * Verify webhook signature.
 * Adjust this to match your payment provider's signature scheme.
 * Expected headers:
 * - x-webhook-signature: HMAC SHA256 hex signature of raw body
 */
function verifySignature(rawBody, signatureHeader) {
  if (!WEBHOOK_SECRET) return false;
  if (!signatureHeader) return false;

  const expected = crypto
    .createHmac('sha256', WEBHOOK_SECRET)
    .update(rawBody)
    .digest('hex');

  try {
    return crypto.timingSafeEqual(Buffer.from(expected), Buffer.from(signatureHeader));
  } catch {
    return false;
  }
}

router.post('/webhooks/payment', express.raw({ type: 'application/json' }), async (req, res) => {
  try {
    const signature = req.get('x-webhook-signature');
    const rawBody = req.body;

    if (!Buffer.isBuffer(rawBody)) {
      return res.status(400).json({ error: 'Invalid raw body' });
    }

    if (!verifySignature(rawBody, signature)) {
      return res.status(401).json({ error: 'Invalid signature' });
    }

    let event;
    try {
      event = JSON.parse(rawBody.toString('utf8'));
    } catch {
      return res.status(400).json({ error: 'Invalid JSON payload' });
    }

    // Example expected structure:
    // {
    //   "type": "payment.succeeded",
    //   "data": {
    //     "orderId": "123",
    //     "paymentStatus": "paid"
    //   }
    // }

    const eventType = event?.type;
    const orderId = event?.data?.orderId;

    if (!eventType || !orderId) {
      return res.status(400).json({ error: 'Missing event type or orderId' });
    }

    let newStatus;

    switch (eventType) {
      case 'payment.succeeded':
        newStatus = 'paid';
        break;
      case 'payment.failed':
        newStatus = 'payment_failed';
        break;
      case 'payment.refunded':
        newStatus = 'refunded';
        break;
      case 'payment.cancelled':
        newStatus = 'cancelled';
        break;
      default:
        // Ignore unhandled events but acknowledge receipt
        return res.status(200).json({ received: true, ignored: true });
    }

    await updateOrderStatus(orderId, newStatus, event);

    return res.status(200).json({ received: true, updated: true });
  } catch (err) {
    console.error('Webhook processing error:', err);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;